app/modules/net_info.c

```diff
@@ -0,0 +1,176 @@
+/* 
+ * http://blog.mclemon.io/esp8266-contributing-to-the-nodemcu-ecosystem
+ * https://github.com/smcl/nodemcu-firmware/blob/cc04aaf92c1c076c30ef0b0eee43b3f924137440/app/modules/test.c
+ *
+ * extracted ping function
+ * adoptet jan 2017 for commit to nodeMCU by Wolfgang Rosner
+ */
+
+
+#include "module.h"
+#include "lauxlib.h"
+#include "platform.h"
+
+#include "c_stdlib.h"
+
+#include "lwip/ip_addr.h"
+#include "espconn.h"
+#include "lwip/dns.h" 
+#include "lwip/app/ping.h"
+#include "lwip/raw.h"
+#include "c_stdio.h"
+
+
+
+static lua_State *gL = NULL;
+static int ping_callback_ref; 
+static int ping_host_count;
+static ip_addr_t ping_host_ip;
+
+
+void ping_received(void *arg, void *data) {
+    // this would require change of the interface 
+    // struct ping_msg *pingmsg = (struct ping_msg*)arg;
+    // struct ping_option *pingopt = pingmsg->ping_opt;
+    struct ping_option *pingopt = (struct ping_option*)arg;
+    struct ping_msg *pingmsg = pingopt->parent_msg ; 
+
+    struct ping_resp *pingresp = (struct ping_resp*)data;
+
+    char ipaddrstr[16];
+    ip_addr_t source_ip;
+    
+    source_ip.addr = pingopt->ip;
+    ipaddr_ntoa_r(&source_ip, ipaddrstr, sizeof(ipaddrstr));
+
+    if (ping_callback_ref != LUA_NOREF) {
+      lua_rawgeti(gL, LUA_REGISTRYINDEX, ping_callback_ref);
+      lua_pushinteger(gL, pingresp->bytes);
+      lua_pushstring(gL, ipaddrstr);
+      lua_pushinteger(gL, pingresp->seqno);
+      lua_pushinteger(gL, pingresp->ttl);
+      lua_pushinteger(gL, pingresp->resp_time);
+      lua_call(gL, 5, 0);
+    } else {
+      c_printf("%d bytes from %s, icmp_seq=%d ttl=%d time=%dms\n",
+	       pingresp->bytes,
+	       ipaddrstr,
+	       pingresp->seqno,
+	       pingresp->ttl,
+	       pingresp->resp_time);
+    }
+}
+
+
+
+static void ping_by_hostname(const char *name, ip_addr_t *ipaddr, void *arg) {
+    struct ping_option *ping_opt = (struct ping_option *)c_zalloc(sizeof(struct ping_option));
+
+    if (ipaddr == NULL) {
+      c_printf("SEVERE problem resolving hostname - network and DNS accessible?\n");
```

Comment:
@devsaurus , @marcelstoer 
> This and below: NODE_ERR() instead of c_printf()?

> . and if you don't want to use NODE_ERR directly you can stick to https://github.com/nodemcu/nodemcu-firmware/wiki/%5BDRAFT%5D-How-to-write-a-C-module#debug-and-error-messages

When I get this right, you want to have output dependent on some debug flag switched on.
However, I think that does not meet the expected typical use case of such network utility tools:

Somebody uses MASTER (presumably from build service) and has some network stuff enabled (both wireless and maybe once a day wired as well). He does **not switch any debug stuff on,** because he expect the **modules on nodemcu level working** without trouble. But he may still want to **debug his own local network** configuration - which is the standard use case of matured network utils like ping, traceroute or ifconfig.

Eg. having no hostname resolution because WIFI is down or DNS is not configured is not an issue for debugging nodemcu code, is it? Notifying the user that host resolution does not work is a perfect expectable behaviour of a ping tool - no need to debug the tool

Well, yes, the two tierd approach of "problem..." and "SEVERE problem..." is somewhat odd and presumably due to the fact that the template resides on the quirky deprecated `espconn` obscuration layer.
I just realized that there are cases (when wifi is down, e.g.) where we don't even have a ipaddr object, and there are cases where we have such, but the hostname could not be found.
Maybe, for the moment I rephrase it in "host name not found" and "name resolution not working"?

So we are talking about **different layers of debugging targets,** don't we?
So the **output of the net util tools has to go to stdout, not to stderr,** or it's conceptual counterparts in nodemcu and lua land.
Please correct me if I'm wrong.